pilopt/src/lib.rs

```diff
@@ -465,6 +466,52 @@ fn remove_constant_witness_columns<T: FieldElement>(pil_file: &mut Analyzed<T>)
     substitute_polynomial_references(pil_file, constant_polys);
 }
 
+/// Identifies intermediate columns that are constrained to a single value, replaces every
+/// reference to this column by the value and deletes the column.
+/// This assumes that intermediate columns are defined before they are used in source order.
+fn remove_constant_intermediate_columns<T: FieldElement>(pil_file: &mut Analyzed<T>) {
+    let constant_intermediate_polys = pil_file
+        .intermediate_polys_in_source_order()
+        .filter(|(symbol, _)| !symbol.is_array())
+        .fold(
+            BTreeMap::default(),
+            |mut values: BTreeMap<PolyID, (String, BigUint)>, (symbol, definitions)| {
+                let ((name, poly_id), mut definition) = symbol
+                    .array_elements()
+                    .zip_eq(definitions.iter().cloned())
+                    .next()
+                    .unwrap();
+                // replace constant intermediate columns by their value
+                definition.post_visit_expressions_mut(&mut |e| {
+                    if let AlgebraicExpression::Reference(AlgebraicReference { poly_id, .. }) = e {
+                        if let Some((_, value)) = values.get(poly_id) {
+                            *e = AlgebraicExpression::Number(T::from(value.clone()));
+                        }
+                    }
+                });
+
+                // simplify the expression
+                let definition = simplify_expression(definition.clone());
```

Comment:
Should we really iteratively replace the values right away or rather keep the steps simple and run them multiple times? Do we have that many intermediate columns that reference other intermediate columns?

Reply:
The use case here is that when we build the bus accumulator updates:

```rust
    col intermediate_fingerprint_0 = 0;
    col intermediate_fingerprint_1 = 0;
    col intermediate_fingerprint_0_1 = std::prelude::challenge(0, 1) * main::intermediate_fingerprint_0 + 11 * std::prelude::challenge(0, 2) * main::intermediate_fingerprint_1 + 0;
    col intermediate_fingerprint_1_1 = std::prelude::challenge(0, 2) * main::intermediate_fingerprint_0 + std::prelude::challenge(0, 1) * main::intermediate_fingerprint_1 + 0;
```

All of this should be removed as it's all 0.

We have a lot of intermediates which depend on each other (depending on the number of columns we're sending/receiving), but they typically don't end up being constant. This only happens in some special cases like above.

I *think* running it twice without memory would work, but writing the case with memory is not that much code. Up to you :)